File: src/main/java/org/gautelis/durga/plugins/JsonTransform.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/**
 * Applies a dot-notation expression to remap fields from input to output.
 *
 * <p>Expression syntax:
 * <pre>
 * "field1, field2, nested.field3"
 *   → copies those fields from input to output at the same paths
 *
 * "field1:dest1, nested.field2:outer.dest2"
 *   → copies source field to destination path
 *
 * "key:value" (with no input field)
 *   → sets a literal value (numbers auto-detected)
 * </pre>
 */
public final class JsonTransform implements Plugin {

    @Override
    public String execute(String payload, String config) throws Exception {
        return transform(payload, config);
    }

    private JsonTransform() {
    }

    /**
     * Transforms input JSON using a mapping expression.
     *
     * @param json       input JSON string
     * @param expression comma-separated field mappings
     * @return transformed JSON string
     * @throws IllegalArgumentException if the expression or JSON is malformed
     */
    public static String transform(String json, String expression) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }
        if (expression == null || expression.isBlank() || expression.equals(".")) {
            return json;
        }

        ObjectNode output = mapper.createObjectNode();
        for (String mapping : expression.split(",")) {
            mapping = mapping.trim();
            if (mapping.isEmpty()) {
                continue;
            }
            int colonIdx = mapping.indexOf(':');
            if (colonIdx >= 0) {
                String left = mapping.substring(0, colonIdx).trim();
                String right = mapping.substring(colonIdx + 1).trim();
                JsonNode sourceValue = PipelinePlugin.fieldAt(input, left);
                if (sourceValue != null) {
                    PipelinePlugin.setFieldAt(output, right, sourceValue);
                } else {
                    JsonNode literalValue = tryParseLiteral(right);
                    if (literalValue != null) {
                        PipelinePlugin.setFieldAt(output, left, literalValue);
                    }
                }
            } else {
                JsonNode sourceValue = PipelinePlugin.fieldAt(input, mapping);
                if (sourceValue != null) {
                    PipelinePlugin.setFieldAt(output, mapping, sourceValue);
                }
            }
        }
        return output.toString();
    }

    private static JsonNode tryParseLiteral(String value) {
        ObjectMapper mapper = new ObjectMapper();
        if ("true".equals(value)) {
            return mapper.getNodeFactory().booleanNode(true);
        }
        if ("false".equals(value)) {
            return mapper.getNodeFactory().booleanNode(false);
        }
        if ("null".equals(value)) {
            return mapper.getNodeFactory().nullNode();
        }
        try {
            long l = Long.parseLong(value);
            return mapper.getNodeFactory().numberNode(l);
        } catch (NumberFormatException e1) {
            try {
                double d = Double.parseDouble(value);
                return mapper.getNodeFactory().numberNode(d);
            } catch (NumberFormatException e2) {
                return mapper.getNodeFactory().textNode(value);
            }
        }
    }
}

File: durga-rust/src/plugins/json_transform.rs
//! Port of `org.gautelis.durga.plugins.JsonTransform`.
//!
//! Remaps fields from input to output using a comma-separated expression:
//! * `"field1, nested.field2"` copies those fields at the same paths;
//! * `"src:dest"` copies a source field to a destination path;
//! * `"key:value"` (source absent) sets a literal value (number/bool/null
//!   auto-detected).

use serde_json::Value;

use crate::pipeline::{field_at, set_field_at};
use crate::plugin::{Plugin, PluginError};

#[derive(Default)]
pub struct JsonTransform;

impl JsonTransform {
    pub fn new() -> Self {
        JsonTransform
    }

    /// Transforms `json` using the mapping `expression`.
    pub fn transform(json: &str, expression: &str) -> Result<String, PluginError> {
        let input: Value = serde_json::from_str(json).map_err(|e| -> PluginError {
            format!("Invalid input JSON: {e}").into()
        })?;

        let trimmed = expression.trim();
        if trimmed.is_empty() || trimmed == "." {
            return Ok(json.to_string());
        }

        let mut output = Value::Object(serde_json::Map::new());
        for raw in expression.split(',') {
            let mapping = raw.trim();
            if mapping.is_empty() {
                continue;
            }
            match mapping.find(':') {
                Some(idx) => {
                    let left = mapping[..idx].trim();
                    let right = mapping[idx + 1..].trim();
                    if let Some(source) = field_at(&input, left) {
                        set_field_at(&mut output, right, source.clone());
                    } else {
                        set_field_at(&mut output, left, parse_literal(right));
                    }
                }
                None => {
                    if let Some(source) = field_at(&input, mapping) {
                        set_field_at(&mut output, mapping, source.clone());
                    }
                }
            }
        }
        Ok(output.to_string())
    }
}

fn parse_literal(value: &str) -> Value {
    match value {
        "true" => Value::Bool(true),
        "false" => Value::Bool(false),
        "null" => Value::Null,
        _ => {
            if let Ok(i) = value.parse::<i64>() {
                Value::from(i)
            } else if let Ok(f) = value.parse::<f64>() {
                Value::from(f)
            } else {
                Value::String(value.to_string())
            }
        }
    }
}

impl Plugin for JsonTransform {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        Ok(Some(Self::transform(payload, config)?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn copies_and_renames_fields() {
        let out = JsonTransform::transform(
            r#"{"id":7,"customer":{"name":"Alice","email":"a@x.com"},"total":9.5,"status":"pending"}"#,
            "id:order_id, customer.name:customer_name, customer.email:customer_email, total:amount, status",
        )
        .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["order_id"], 7);
        assert_eq!(v["customer_name"], "Alice");
        assert_eq!(v["customer_email"], "a@x.com");
        assert_eq!(v["amount"], 9.5);
        assert_eq!(v["status"], "pending");
    }

    #[test]
    fn sets_literal_when_source_absent() {
        let out = JsonTransform::transform(r#"{"a":1}"#, "missing:42, flag:true").unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["missing"], 42);
        assert_eq!(v["flag"], true);
    }

    #[test]
    fn identity_expression_passes_through() {
        let json = r#"{"a":1}"#;
        assert_eq!(JsonTransform::transform(json, ".").unwrap(), json);
    }
}
